core/views.py

```diff
@@ -1,16 +1,22 @@
 from django.shortcuts import render
+from django.db.models.functions import TruncDay
+from django.db.models import Count
+from collections import defaultdict
 
 from .models import Event, Venue
 
 
 def index(request):
-    all_events = Event.objects.all()
+    all_events = Event.objects.all().order_by("-starttime").annotate(day=TruncDay("starttime"))
+    grouped_events = defaultdict(list)
+    for event in all_events:
+        grouped_events[event.day].append(event)
     all_venues = Venue.objects.all()
     return render(
         request,
         "core/index.html",
         {
-            "all_events": all_events,
+            "all_events": dict(grouped_events),
```

Comment:
I'd be very inclined to just pass sorted(grouped_events.items) to the template — having the extra dict wrapper makes it harder to understand what all is going on in my opinion, and makes me less certain that the dates are going to be displayed in order

Reply:
That's not exactly the same though. Here, the dict's keys are the dates, and the .items are, for each date, that date's events.

Passing the .items would only pass the events - ungrouped :) 

Does it make sense? Happy to entertain other solutions of course.

Reply:
As I understand it grouped_events.items() should give you like

[(date1, [eventondate1-1, eventondatae1-2,...]), (date2, [eventondate2-1, eventondate2-2, ...]), ...]

so, grouped by date. (I still think we should construct grouped_events to make that grouping!!!)

it's grouped_events.values() that'll only give you the events.

Reply:
Yes .items() gives a type of [(date1, [events]), (date2, [events])]. I tried the change and it works, however i think python keeps the insertion order in dicts, so they should be displayed as intended

Reply:
If python guarantees preserved insertion order this is fine because of the sort provided by the query! Thanks for satisfying my curiosity there.

I still think it's slightly simpler to just send the grouped_events to the template - I don't think the coercion to dict() does anything helpful

```suggestion
            "all_events": grouped_events,
```

Reply:
@gregsadetsky mentioned that defaultdict can sometimes act weird in django templates, which is why line 19 is like that. i can either move the dict() conversion right after creating grouped_events or i can pass in grouped_events.items(), whichever people think is easier to understand

Reply:
Nah if there's some deep lore reasons for the dict() conversion stick with it! It felt weird to me but it's very good to know there's a reason.

Reply:
Ok super, let's keep the conversion -- @andrea749 if you don't mind, maybe you could add a comment right above the `dict()` conversion that points to https://stackoverflow.com/a/64666307 ? that's the exact same problem/solution as we're doing here. it will make it clearer in the future for anyone wondering why we're doing that

thank you!!